### PR TITLE
Better unit tangentials data structure

### PR DESCRIPTION
**Pull request by bangerth**
#2424 points out that we could use better data structures for the `Mapping*::InternalData::unit_tangentials` fields. This patch implements this for `MappingQGeneric` and `MappingFEField`. The current patch for `MappingManifold` in #2272 still uses the old scheme and would need to eventually be converted.

In reference to #1198.

### CODE REVIEW

File: include/deal.II/fe/mapping_fe_field.h [hunk 1]
```diff
@@ -361,7 +363,7 @@ class MappingFEField : public Mapping<dim,spacedim>
      * Unit tangential vectors. Used for the computation of boundary forms and
      * normal vectors.
      *
-     * This vector has (dim-1)GeometryInfo::faces_per_cell entries. The first
+     * This vector has (dim-1)*GeometryInfo::faces_per_cell entries. The first
```
Comment (drwells) [inline include/deal.II/fe/mapping_fe_field.h]: Shouldn't it say 'This array'?

File: include/deal.II/fe/mapping_q_generic.h [hunk 2]
```diff
@@ -360,7 +362,7 @@ class MappingQGeneric : public Mapping<dim,spacedim>
      * Unit tangential vectors. Used for the computation of boundary forms and
      * normal vectors.
      *
-     * This vector has (dim-1)GeometryInfo::faces_per_cell entries. The first
+     * This vector has (dim-1)*GeometryInfo::faces_per_cell entries. The first
```
Comment (drwells) [inline include/deal.II/fe/mapping_q_generic.h]: same

File: source/fe/mapping_q_generic.cc [hunk 3]
```diff
@@ -764,32 +764,30 @@ initialize_face (const UpdateFlags      update_flags,
         {
           aux.resize (dim-1, std::vector<Tensor<1,spacedim> > (n_original_q_points));
 
-          // Compute tangentials to the
-          // unit cell.
-          const unsigned int nfaces = GeometryInfo<dim>::faces_per_cell;
-          unit_tangentials.resize (nfaces*(dim-1),
-                                   std::vector<Tensor<1,dim> > (n_original_q_points));
+          // Compute tangentials to the unit cell.
+          for (unsigned int i=0; i!=GeometryInfo<dim>::faces_per_cell * (dim-1); ++i)
```
Comment (drwells) [inline source/fe/mapping_q_generic.cc]: While they are equivalent, I think `unit_tangentials.size()` would be a little nicer here.

**Comment by drwells** [comment]
Small comments aside this is a very nice patch and passes (large parts of) the test suite.

**Comment by drwells** [comment]
A clarification that struck me as I walked home: I only reran tests in `fe`, `base`, and `bits` and they all passed (I did not mean to imply that any tests failed).

**Comment by bangerth** [comment]
Comments addressed.

For reference, I don't think that you should feel morally obliged to run the testsuite for every PR you look at. Doing so is the job of the author of the patch (or, once we get to implement that, of a continuous integration server). If something breaks due to a patch, then that's on the author, not the reviewer.

**Comment by drwells** [comment]
I have access to a 16-core machine (that is usually not doing other things) that can run the whole test suite in, very roughly, an hour. I understand what you are saying but it is very little work for me to compile and run things since I have decent hardware (thanks, RTG!) and it makes me feel better about merging (and does ocassionally catch stuff).

**Pull request merged**